### PR TITLE
raise BrokenResourceError if lock owner has exited.

### PR DESCRIPTION
**Pull request by jakkdl**
Partially addresses #3035 

* Idk if adding `test_lock_multiple_acquire` is very meaningful
* need to add newsfragment

It's possible that a full implementation for #3035 makes this redundant, but this was easy enough to write&test I thought I might as well push it as a PR that improves the status quo for now.

### CODE REVIEW

File: src/trio/_tests/test_sync.py [hunk 1]
```diff
@@ -586,3 +587,31 @@ async def lock_taker() -> None:
...
+        await lock.acquire()
+
+
+@slow
+async def test_lock_multiple_acquire() -> None:
+    """This currently hangs without throwing any error
+    See https://github.com/python-trio/trio/issues/3035
+    """
+    lock = trio.Lock()
+    with pytest.raises(trio.TooSlowError):
```
Comment (A5rocks) [inline src/trio/_tests/test_sync.py:612]: How about xfailing this test?

File: src/trio/_sync.py [hunk 2]
```diff
@@ -571,11 +572,19 @@ def acquire_nowait(self) -> None:
...
             raise RuntimeError("attempt to re-acquire an already held Lock")
         elif self._owner is None and not self._lot:
             # No-one owns it
             self._owner = task
+        elif (
+            self._owner is not None
+            and inspect.getcoroutinestate(self._owner.coro) is inspect.CORO_CLOSED
+        ):
+            raise trio.BrokenResourceError(
+                "Attempted acquire on Lock which will never be released, owner has exited.",
```
Comment (A5rocks) [inline src/trio/_sync.py:586]:
```suggestion
                "Attempted acquire on Lock which will never be released as the owner has exited.",
```

I'm not sure this is the ideal message either way but the alternative I can think of isn't great either: "Deadlock detected: acquiring lock which will never be released."

**Review by A5rocks — CHANGES_REQUESTED** [review]
Can you add a newsfragment?

**Comment by A5rocks** [comment]
> I'm somewhat concerned about the performance impact of `inspect.getcoroutinestate` in `acquire_nowait()`, which might be fine.

It should only be in the case where we would lock anyways.

> The larger problem is that it only addresses the case where the holding task is _already_ gone at acquire-time; I think we should probably go directly to the general solution of adding a `ParkingLot.mark_resource_broken()` method and using that from Lock.

Yeah, the problem is we can't just add a way to poison a `ParkingLot`, we also need a way to know a `Lock`'s task got ended. We can't use a `weakref.finalizer` because no guarantees on when `Task` gets deleted (it could be stored somewhere else other than `_owner`), so I think we need something in `trio`'s event loop itself.

**Comment by jakkdl** [comment]
> I'm somewhat concerned about the performance impact of `inspect.getcoroutinestate` in `acquire_nowait()`, which might be fine.

I ran a simple benchmark now and could not detect any difference with or without a call to `inspect.getcoroutinestate`.

> The larger problem is that it only addresses the case where the holding task is _already_ gone at acquire-time; I think we should probably go directly to the general solution of adding a `ParkingLot.mark_resource_broken()` method and using that from Lock.

Yes this PR intentionally only made the easy fix from https://github.com/python-trio/trio/issues/3035#issuecomment-2312328838, but if the full fix turns out doable then we'll close this one.

**Comment by jakkdl** [comment]
This will probably end up closed in favor of #3081

**Comment by A5rocks** [comment]
Now unnecessary.

**Pull request closed without merging**